[stdlib] AutoreleasingUnsafeMutablePointer: eliminate questionable pointer use

lorentey opened the pull request:
`AutoreleasingUnsafeMutablePointer` is pointing to a +0 reference, but in its `pointee` property’s getter/setter implementations, it is loading the pointer into regular `Unsafe[Mutable]Pointer`s. Those are assuming that the addressed memory contains a +1 reference, which can mislead the compiler into doing optimizations that aren’t valid.

Change the getter/setter implementations so that they use `UnsafeRawPointer` and load/store `Unmanaged` values instead. As long as `Unmanaged.passUnretained(_:)` and `Unmanaged.takeUnretainedValue()` do the right thing, then `AutoreleasingUnsafeMutablePointer` won’t have issues, either. (This boils down to ensuring that loading a strong reference out of an `unmanaged(unsafe)` value works correctly.)

lorentey committed:
[stdlib] AutoreleasingUnsafeMutablePointer: eliminate questionable pointer use

AutoreleasingUnsafeMutablePointer is pointing to a +0 reference, but in its pointee property’s getter/setter implementations, it is loading the pointer into regular Unsafe[Mutable]Pointers. Those are assuming that the addressed memory contain a +1 reference, which can mislead the compiler into doing optimizations that aren’t valid.

Change the getter/setter implementations so that they use UnsafeRawPointer and load/store Unmanaged values instead. As long as Unmanaged.passUnretained(_:) and Unmanaged.takeUnretainedValue() do the right thing, then AutoreleasingUnsafeMutablePointer won’t have issues, either. (This boils down to ensuring that loading a strong reference out of an unmanaged(unsafe) value works correctly.)

lorentey commented:
This is still somewhat of a WIP - AutoreleasingUnsafeMutablePointer doesn't seem to have much test coverage yet, so perhaps this is the right PR to add those.

lorentey commented:
@swift-ci benchmark

Catfish-Man commented on stdlib/public/core/BridgeObjectiveC.swift:
```diff
@@ -376,9 +376,22 @@ public struct AutoreleasingUnsafeMutablePointer<Pointee /* TODO : class */>
+      // The memory addressed by this pointer contains a non-owning reference,
+      // therefore we *must not* point an `UnsafePointer<AnyObject>` to
+      // it---otherwise we would allow the compiler to assume it has a +1
+      // refcount, enabling some optimizations that wouldn't be valid.
```
This makes sense but is kinda terrifying. I need to think carefully about some of the NSDictionary bridging stuff…

lorentey replied:
All +0 references in memory are a hazard; they must only be handled through the `Unmanaged` glovebox.  E.g., we'll need to audit any and all NSFastEnumeration implementations & usages from Swift code.

(Currently Unmanaged doesn't quite get this right either, but Michael is already working on fixing that.)

lorentey commented:
The benchmark results seem irrelevant; I don't see how any of these benchmarks would touch `AutoreleasingUnsafeMutablePointer` paths.

However, this new version does generate a little bit worse code:

```swift
// SOURCE:
class Foo {}

func foo(_ p: AutoreleasingUnsafeMutablePointer<Foo>) -> Foo {
  return p.pointee
}
```

```swift
// ORIGINAL SIL:
sil hidden [signature_optimized_thunk] [always_inline] @$s6output3fooyAA3FooCSAyADGF : $@convention(thin) (AutoreleasingUnsafeMutablePointer<Foo>) -> @owned Foo {
bb0(%0 : $AutoreleasingUnsafeMutablePointer<Foo>):
  debug_value %0 : $AutoreleasingUnsafeMutablePointer<Foo>, let, name "p", argno 1 // id: %1
  %2 = struct_extract %0 : $AutoreleasingUnsafeMutablePointer<Foo>, #AutoreleasingUnsafeMutablePointer._rawValue // user: %3
  %3 = pointer_to_address %2 : $Builtin.RawPointer to [strict] $*Foo // user: %4
  %4 = load %3 : $*Foo                            // users: %5, %6
  strong_retain %4 : $Foo                         // id: %5
  return %4 : $Foo                                // id: %6
} // end sil function '$s6output3fooyAA3FooCSAyADGF'

// NEW SIL:
sil hidden @$s6output3fooyAA3FooCSAyADGF : $@convention(thin) (AutoreleasingUnsafeMutablePointer<Foo>) -> @owned Foo {
bb0(%0 : $AutoreleasingUnsafeMutablePointer<Foo>):
  debug_value %0 : $AutoreleasingUnsafeMutablePointer<Foo>, let, name "p", argno 1 // id: %1
  %2 = struct_extract %0 : $AutoreleasingUnsafeMutablePointer<Foo>, #AutoreleasingUnsafeMutablePointer._rawValue // user: %3
  %3 = pointer_to_address %2 : $Builtin.RawPointer to $*Optional<Unmanaged<AnyObject>> // user: %4
  %4 = load %3 : $*Optional<Unmanaged<AnyObject>> // user: %5
  switch_enum %4 : $Optional<Unmanaged<AnyObject>>, case #Optional.some!enumelt.1: bb1, case #Optional.none!enumelt: bb2 // id: %5

bb1(%6 : $Unmanaged<AnyObject>):                  // Preds: bb0
  %7 = struct_extract %6 : $Unmanaged<AnyObject>, #Unmanaged._value // user: %8
  %8 = unmanaged_to_ref %7 : $@sil_unmanaged AnyObject to $AnyObject // users: %10, %9
  %9 = enum $Optional<AnyObject>, #Optional.some!enumelt.1, %8 : $AnyObject // user: %11
  strong_retain %8 : $AnyObject                   // id: %10
  br bb3(%9 : $Optional<AnyObject>)               // id: %11

bb2:                                              // Preds: bb0
  %12 = enum $Optional<AnyObject>, #Optional.none!enumelt // user: %13
  br bb3(%12 : $Optional<AnyObject>)              // id: %13

bb3(%14 : $Optional<AnyObject>):                  // Preds: bb2 bb1
  %15 = metatype $@thick Optional<AnyObject>.Type // user: %16
  %16 = builtin "sizeof"<Optional<AnyObject>>(%15 : $@thick Optional<AnyObject>.Type) : $Builtin.Word // user: %17
  %17 = builtin "sextOrBitCast_Word_Int64"(%16 : $Builtin.Word) : $Builtin.Int64 // user: %21
  %18 = metatype $@thick Foo.Type                 // user: %19
  %19 = builtin "sizeof"<Foo>(%18 : $@thick Foo.Type) : $Builtin.Word // user: %20
  %20 = builtin "sextOrBitCast_Word_Int64"(%19 : $Builtin.Word) : $Builtin.Int64 // user: %21
  %21 = builtin "cmp_eq_Int64"(%17 : $Builtin.Int64, %20 : $Builtin.Int64) : $Builtin.Int1 // user: %23
  %22 = integer_literal $Builtin.Int1, -1         // user: %23
  %23 = builtin "xor_Int1"(%21 : $Builtin.Int1, %22 : $Builtin.Int1) : $Builtin.Int1 // user: %24
  cond_fail %23 : $Builtin.Int1, "Can't unsafeBitCast between types of different sizes" // id: %24
  %25 = unchecked_ref_cast %14 : $Optional<AnyObject> to $Foo // user: %26
  return %25 : $Foo                               // id: %26
} // end sil function '$s6output3fooyAA3FooCSAyADGF'
```

```asm
// Original x86_64 assembly:
output.foo(Swift.AutoreleasingUnsafeMutablePointer<output.Foo>) -> output.Foo:
        push    rbp
        mov     rbp, rsp
        mov     rdi, qword ptr [rdi]
        pop     rbp
        jmp     _swift_retain

// New x86_64 assembly:
output.foo(Swift.AutoreleasingUnsafeMutablePointer<output.Foo>) -> output.Foo:
        push    rbp
        mov     rbp, rsp
        push    rbx
        push    rax
        mov     rbx, qword ptr [rdi]
        test    rbx, rbx
        je      LBB5_1
        mov     rdi, rbx
        call    _swift_unknownObjectRetain
        jmp     LBB5_3
LBB5_1:
        xor     ebx, ebx
LBB5_3:
        mov     rax, rbx
        add     rsp, 8
        pop     rbx
        pop     rbp
        ret
```

`AutoreleasingUnsafeMutablePointer` isn't particularly performance sensitive, though.

lorentey committed:
[stdlib] AutoreleasingUnsafeMutablePointer: update docs

lorentey committed:
[stdlib] Make Unmanaged.takeUnretainedValue, .takeRetainedValue transparent

This improves codegen in debug builds.

lorentey committed:
[stdlib] AutoreleasingUnsafeMutablePointer: use typed pointers rather than UnsafeRawPointer

This improves code generation in -Onone mode. (UnsafeRawPointer.load/.storeBytes isn’t @_transparent and it doesn’t get specialized in unoptimized builds.)

lorentey commented:
The latest commits get rid of unspecialized generic calls (to `UnsafeRawPointer.load`/`.storeBytes` and `Unmanaged.takeUnretainedValue`) in `-Onone` builds.

lorentey commented:
(And they also use my new favorite type, `UnsafePointer<Optional<Unmanaged<AnyObject>>>`.)

atrick commented:
@lorentey thanks for figuring out the semantics and documenting it. Just to limit the number of times `unsafeBitCast` appears in the code--since each occurrence is generally considered a "bug"--you can use `_unsafeReferenceCast` as long as the values are dynamically `Optional<AnyObject>`.

lorentey committed:
[stdlib] AutoreleasingMutableUnsafePointer: switch to _unsafeReferenceCast for extra validation

lorentey commented:
@swift-ci test

lorentey committed:
[test] Add some high-level tests for AutoreleasingUnsafeMutablePointer
